From 4f71f65e4b49128a27e54be07f7740a17913c187 Mon Sep 17 00:00:00 2001
From: Christian Balster <christian.balster@gmail.com>
Date: Sat, 22 Nov 2014 17:25:31 +0100
Subject: [PATCH 2/3] [1/2] Settings: add option to block camera access

Adds an option to disable the front and/or back cameras of the device.
This is implemented similar to the way it is done for DevicePolicies so
there should be no (additional) sideeffects.

Change-Id: I715f1762e76b1b53bea7dad32b233fde59bef517
---
 res/values/custom_arrays.xml                   | 22 +++++++++++++++++
 res/values/custom_strings.xml                  | 12 ++++++++++
 res/xml/security_settings_misc.xml             | 11 +++++++++
 src/com/android/settings/SecuritySettings.java | 33 ++++++++++++++++++++++++++
 4 files changed, 78 insertions(+)

diff --git a/res/values/custom_arrays.xml b/res/values/custom_arrays.xml
index 11c7286a8..f4205bc8d 100644
--- a/res/values/custom_arrays.xml
+++ b/res/values/custom_arrays.xml
@@ -297,4 +297,26 @@
         <item>0</item>
         <item>1</item>
     </string-array>
+
+    <!-- Disable Camera Option -->
+    <string-array name="disable_cameras_entries" translatable="false">
+        <item>@string/disable_cameras_entries_none</item>
+        <item>@string/disable_cameras_entries_front</item>
+        <item>@string/disable_cameras_entries_back</item>
+        <item>@string/disable_cameras_entries_all</item>
+    </string-array>
+
+    <string-array name="disable_cameras_values" translatable ="false">
+        <item>0</item>
+        <item>1</item>
+        <item>2</item>
+        <item>3</item>
+    </string-array>
+
+    <string-array name="disable_cameras_summaries" translatable="false">
+        <item>@string/disable_cameras_summary_none</item>
+        <item>@string/disable_cameras_summary_front</item>
+        <item>@string/disable_cameras_summary_back</item>
+        <item>@string/disable_cameras_summary_all</item>
+    </string-array>
 </resources>
diff --git a/res/values/custom_strings.xml b/res/values/custom_strings.xml
index efc89669e..4cacfef03 100644
--- a/res/values/custom_strings.xml
+++ b/res/values/custom_strings.xml
@@ -384,7 +384,19 @@
     <string name="headset_plug_force_actions_title">Force remove action</string>
     <string name="headset_plug_force_actions_summary">Run remove action also if shortcut was already running</string>
 
+
     <!-- Development shortcut -->
     <string name="development_shortcut_title">Development shortcut</string>
     <string name="development_shortcut_summary">Enable \'wipe\' and \'force close\' options in notification area and recents app list</string>
+    <!-- Disable Camera Option -->
+    <string name="hardware_title">Hardware</string>
+    <string name="disable_cameras_title">Disable Device Cameras</string>
+    <string name="disable_cameras_summary_none">Cameras are enabled</string>
+    <string name="disable_cameras_summary_front">Front facing camera is disabled</string>
+    <string name="disable_cameras_summary_back">Back facing camera is disabled</string>
+    <string name="disable_cameras_summary_all">All cameras are disabled</string>
+    <string name="disable_cameras_entries_none">None</string>
+    <string name="disable_cameras_entries_front">Front</string>
+    <string name="disable_cameras_entries_back">Back</string>
+    <string name="disable_cameras_entries_all">All</string>
 </resources>
diff --git a/res/xml/security_settings_misc.xml b/res/xml/security_settings_misc.xml
index dbbeb1df4..9d94f8f10 100644
--- a/res/xml/security_settings_misc.xml
+++ b/res/xml/security_settings_misc.xml
@@ -36,6 +36,17 @@
                 android:persistent="false"/>
     </PreferenceCategory>
 
+    <PreferenceCategory android:title="@string/hardware_title"
+                android:persistent="false">
+
+        <ListPreference android:key="disable_cameras"
+                android:title="@string/disable_cameras_title"
+                android:summary="@string/disable_cameras_summary_none"
+                android:entries="@array/disable_cameras_entries"
+                android:entryValues="@array/disable_cameras_values"
+                android:persistent="false"/>
+    </PreferenceCategory>
+
     <PreferenceCategory android:key="device_admin_category"
             android:title="@string/device_admin_title"
             android:persistent="false">
diff --git a/src/com/android/settings/SecuritySettings.java b/src/com/android/settings/SecuritySettings.java
index 241322f76..83ba14dac 100644
--- a/src/com/android/settings/SecuritySettings.java
+++ b/src/com/android/settings/SecuritySettings.java
@@ -29,7 +29,9 @@ import android.content.Intent;
 import android.content.pm.PackageManager;
 import android.content.pm.ResolveInfo;
 import android.content.pm.UserInfo;
+import android.hardware.Camera;
 import android.os.Bundle;
+import android.os.SystemProperties;
 import android.os.UserHandle;
 import android.os.UserManager;
 import android.preference.CheckBoxPreference;
@@ -93,6 +95,8 @@ public class SecuritySettings extends RestrictedSettingsFragment
     private static final String MENU_UNLOCK_PREF = "menu_unlock";
     private static final String KEY_APP_SECURITY_CATEGORY = "app_security";
     private static final String KEY_BLACKLIST = "blacklist";
+    private static final String KEY_DISABLE_CAMERAS = "disable_cameras";
+    private static final String PROPERTY_DISABLE_CAMERAS = "persist.sys.disable_cameras";
 
     private PackageManager mPM;
     private DevicePolicyManager mDPM;
@@ -128,6 +132,7 @@ public class SecuritySettings extends RestrictedSettingsFragment
     private ListPreference mLockNumpadRandom;
     private CheckBoxPreference mMenuUnlock;
     private PreferenceScreen mBlacklist;
+    private ListPreference mDisableCameras;
 
     @Override
     public void onCreate(Bundle savedInstanceState) {
@@ -312,6 +317,18 @@ public class SecuritySettings extends RestrictedSettingsFragment
         mShowPassword = (CheckBoxPreference) root.findPreference(KEY_SHOW_PASSWORD);
         mResetCredentials = root.findPreference(KEY_RESET_CREDENTIALS);
 
+        // Disable Cameras
+        mDisableCameras = (ListPreference) root.findPreference(KEY_DISABLE_CAMERAS);
+        if (mDisableCameras != null) {
+            if (Camera.getNumberOfCameras() == 0) {
+                root.removePreference(mDisableCameras);
+                mDisableCameras = null;
+            } else {
+                mDisableCameras.setOnPreferenceChangeListener(this);
+                updateDisableCameras();
+            }
+        }
+
         // Credential storage
         final UserManager um = (UserManager) getActivity().getSystemService(Context.USER_SERVICE);
         mKeyStore = KeyStore.getInstance(); // needs to be initialized for onResume()
@@ -680,6 +697,9 @@ public class SecuritySettings extends RestrictedSettingsFragment
                     Integer.valueOf((String) value));
             mLockNumpadRandom.setValue(String.valueOf(value));
             mLockNumpadRandom.setSummary(mLockNumpadRandom.getEntry());
+        } else if (preference == mDisableCameras) {
+            SystemProperties.set(PROPERTY_DISABLE_CAMERAS, value == null ? "0" : value.toString());
+            updateDisableCameras();
         }
         return true;
     }
@@ -704,4 +724,17 @@ public class SecuritySettings extends RestrictedSettingsFragment
             }
         }
     }
+
+    private void updateDisableCameras() {
+        String currentValue = SystemProperties.get(PROPERTY_DISABLE_CAMERAS, "0");
+        CharSequence[] values = mDisableCameras.getEntryValues();
+        for (int i = 0; i < values.length; i++) {
+            if (currentValue.contentEquals(values[i])) {
+                mDisableCameras.setValueIndex(i);
+                mDisableCameras.setSummary(getResources().getStringArray(R.array.disable_cameras_summaries)[i]);
+                break;
+            }
+        }
+        return;
+    }
 }
-- 
2.14.1

